frontend/src/components/ui/file-input.ts

```diff
@@ -0,0 +1,225 @@
+import { tw } from "@/utils/tailwind";
+
+import "@/components/ui/file-list";
+
+const droppingClass = tw`bg-slate-100`;
+
+/**
+ * Allow attaching one or more files.
+ *
+ * @fires btrix-change
+ * @fires btrix-remove
+ */
+@customElement("btrix-file-input")
+@localized()
+export class FileInput extends FormControl(TailwindElement) {
+  /**
+   * Form control name, if used as a form control
+   */
+  @property({ type: String })
+  name?: string;
+
+  /**
+   * Specify which file types are allowed
+   */
+  @property({ type: String })
+  accept?: HTMLInputElement["accept"];
+
+  /**
+   * Enable selecting more than one file
+   */
+  @property({ type: Boolean })
+  multiple?: HTMLInputElement["multiple"];
+
+  /**
+   * Enable dragging files into drop zone
+   */
+  @property({ type: Boolean })
+  drop = false;
+
+  @state()
+  private files: File[] = [];
+
+  @query("#dropzone")
+  private readonly dropzone?: HTMLElement | null;
+
+  @query("input[type='file']")
+  private readonly input?: HTMLInputElement | null;
+
+  formResetCallback() {
+    this.files = [];
+
+    if (this.input) {
+      this.input.value = "";
+    }
+  }
+
+  protected willUpdate(changedProperties: PropertyValues): void {
+    if (changedProperties.has("files")) {
+      this.syncFormValue();
+    }
+  }
+
+  private syncFormValue() {
+    const formControlName = this.name;
+
+    if (!formControlName) return;
+
+    // `ElementInternals["setFormValue"]` doesn't support `FileList` yet,
+    // construct `FormData` instead
+    const formData = new FormData();
+
+    this.files.forEach((file) => {
+      formData.append(formControlName, file);
+    });
+
+    this.setFormValue(formData);
+  }
+
+  render() {
+    return html`
+      ${this.files.length ? this.renderFiles() : this.renderInput()}
+    `;
+  }
+
+  private readonly renderInput = () => {
+    return html`
+      <label
+        id="dropzone"
+        class=${clsx(
+          tw`cursor-pointer`,
+          this.drop &&
+            tw`flex size-full flex-col items-center justify-center gap-2.5 rounded p-6 text-center outline-dashed outline-1 -outline-offset-1 outline-neutral-400 transition-all hover:outline-primary-400`,
+        )}
+        @drop=${this.drop ? this.onDrop : undefined}
+        @dragover=${this.drop ? this.onDragover : undefined}
+        @dragenter=${this.drop
+          ? () => this.dropzone?.classList.add(droppingClass)
+          : undefined}
+        @dragleave=${this.drop
+          ? () => this.dropzone?.classList.remove(droppingClass)
+          : undefined}
+      >
+        <input
+          class="sr-only"
+          type="file"
+          accept=${ifDefined(this.accept)}
+          ?multiple=${this.multiple}
+          @change=${() => {
+            const files = this.input?.files;
+
+            if (files) {
+              void this.handleChange(files);
+            }
+          }}
+        />
+        <slot @click=${() => this.input?.click()}></slot>
```

Comment:
I believe this is necessary so the click event doesn't also propagate back to the `<label>`
```suggestion
        <slot
          @click=${(e: MouseEvent) => {
            e.preventDefault();
            this.input?.click();
          }}
        ></slot>
```